### PR TITLE
Deep copy strings in safe_structs

### PR DESCRIPTION
**Pull request by mark-lunarg**
Safe structs weren't so safe.  Strings were not previously copied which could cause crashes if the application reclaimed those resources.  Extended deep copying coverage to all strings in Vulkan structs.

Also default-initialized all pointers.

### CODE REVIEW

File: layers/generated/vk_safe_struct.cpp [hunk 1]
```diff
@@ -37,104 +37,124 @@
...
+    enabledExtensionCount(in_struct->enabledExtensionCount)
 {
     pNext = SafePnextCopy(in_struct->pNext);
+    ppEnabledLayerNames = new char *[in_struct->enabledLayerCount];
+    for (uint32_t i = 0; i < enabledLayerCount; ++i) {
+        (const_cast<const char **>(ppEnabledLayerNames))[i] = SafeStringCopy(in_struct->ppEnabledLayerNames[i]);
+    }
+    ppEnabledExtensionNames = new char *[in_struct->enabledExtensionCount];
+    for (uint32_t i = 0; i < enabledExtensionCount; ++i) {
+        (const_cast<const char **>(ppEnabledExtensionNames))[i] = SafeStringCopy(in_struct->ppEnabledExtensionNames[i]);
```
Comment (krOoze) [inline layers/generated/vk_safe_struct.cpp]: Wouldn't it make more sense to not declare it `const char* const*` in the first place?

File: scripts/helper_file_generator.py [hunk 2]
```diff
@@ -1249,9 +1250,27 @@ def GenerateSafeStructSource(self):
...
+                                # Create deep copies of strings
+                                if member.len:
+                                    copy_strings += '    %s = new char *[in_struct->%s];\n' % (member.name, member.len)
+                                    copy_strings += '    for (uint32_t i = 0; i < %s; ++i) {\n' % member.len
+                                    copy_strings += '        (const_cast<const char **>(%s))[i] = SafeStringCopy(in_struct->%s[i]);\n' % (member.name, member.name)
+                                    copy_strings += '    }\n'
+
+                                    destruct_txt += '    if (%s) {\n' % member.name
+                                    destruct_txt += '        for (uint32_t i = 0; i < %s; ++i) {\n' % member.len
+                                    destruct_txt += '            delete %s[i];\n' % member.name
```
Comment (mikes-lunarg) [inline scripts/helper_file_generator.py]: SafeStringCopy uses new[] so this should be delete[]

File: scripts/helper_file_generator.py [hunk 3]
```diff
@@ -1249,9 +1250,27 @@ def GenerateSafeStructSource(self):
...
+                                    destruct_txt += '        }\n'
+                                    destruct_txt += '        delete [] %s;\n' % member.name
+                                    destruct_txt += '    }\n'
+                                else:
+                                    copy_strings += '    %s = SafeStringCopy(in_struct->%s);\n' % (member.name, member.name)
+                                    destruct_txt += '    if (%s) delete %s;\n' % (member.name, member.name)
```
Comment (mikes-lunarg) [inline scripts/helper_file_generator.py]: Same delete[] comment as above

File: layers/generated/vk_safe_struct.cpp (hunk 1 above)
Comment (mark-lunarg) [reply layers/generated/vk_safe_struct.cpp]: Yes, however the policy is that these structs exactly match the ones in the spec.

File: scripts/helper_file_generator.py [hunk 4]
```diff
@@ -872,9 +873,20 @@ def kenum_to_key(raw_key): return re.search('^kVulkanObjectType(.*)', raw_key).g
...
+        # Construct Safe-struct helper functions
+
+        string_copy_proc = '\n\n'
+        string_copy_proc += 'const char *SafeStringCopy(const char *in_string) {\n'
+        string_copy_proc += '    if (nullptr == in_string) return nullptr;\n'
+        string_copy_proc += '    const auto string_size = strlen(in_string);\n'
+        string_copy_proc += '    char* dest = new char[string_size + 1];\n'
+        string_copy_proc += '    strncpy(dest, in_string, string_size);\n'
+        string_copy_proc += "    dest[string_size] = '\\0';\n"
+        string_copy_proc += '    return dest;\n'
```
Comment (krOoze) [inline scripts/helper_file_generator.py]: Why not simply:

```
char* dest = new char[std::strlen(in_string) + 1];
return std::strcpy(dest, in_string);
```

Comment (mark-lunarg) [reply scripts/helper_file_generator.py]: I was being a little paranoid, but experts agree, this is better. 😐

File: scripts/helper_file_generator.py (hunk 2 above)
Comment (mark-lunarg) [reply scripts/helper_file_generator.py]: Done, both cases.  Thanks!

**Pull request merged**